src/tools/cephfs_mirror/SnapReplicator.cc

```diff
@@ -0,0 +1,434 @@
+  // we traverse the new snap only in dirs with (mtime|ctime) > old snap ctime
+  while (!is_error && !dir_queue.empty()) {
+    std::string dir_name = dir_queue.front();
+    dir_queue.pop();
+
+    if (ceph_opendir(m_src_mnt, ("."s + dir_name).c_str(), &new_dirp) < 0) {
+      std::clog << __func__ << ":" << __LINE__ << ": opendir failed for " << ("." + dir_name) << std::endl;
+      is_error = true;
+      break;
+    }
+    ceph_assert(std::string(ceph_readdir(m_src_mnt, new_dirp)->d_name) == ".");
+    ceph_assert(std::string(ceph_readdir(m_src_mnt, new_dirp)->d_name) == "..");
+
+    while (ceph_readdirplus_r(m_src_mnt, new_dirp, &de, &new_stx, CEPH_STATX_BASIC_STATS, AT_SYMLINK_NOFOLLOW, NULL) > 0) {
+      if (!is_first_sync) {
+        std::string old_dentry = m_src_dir      + "/"s +
+                                 m_src_snap_dir + "/"s +
+                                 m_old_snap     +
+                                 (dir_name == "" ? ""s : (dir_name + "/"s)) +
+                                 (dir_name == "" ? "/"s + de.d_name : de.d_name);
+
+        int rv = 0;
+        if ((rv = ceph_statx(m_src_mnt, old_dentry.c_str(), &old_stx, CEPH_STATX_BASIC_STATS, AT_SYMLINK_NOFOLLOW)) == 0) {
+          // found entry in old snap
+          std::clog << __func__ << ":" << __LINE__ << ": processing old entry " << old_dentry << std::endl;
+          if (handle_old_entry(dir_name, old_dentry, de, old_stx, new_stx, old_rctime, dir_queue) < 0) {
+            is_error = true;
+            break;
+          }
+        } else if (rv == -ENOENT) {
+          goto new_entry;
+        } else {
+          std::clog << __func__ << ":" << __LINE__ << ": error while getting statx for old entry " << old_dentry << std::endl;
+          is_error = true;
+          break;
+        }
+      } else {
+new_entry:
+        std::string dentry = dir_name + "/"s + de.d_name;
+        std::string dst_dentry_path = m_src_dir + dentry;
+
+        std::clog << __func__ << ":" << __LINE__ << ": processing new entry " << dst_dentry_path << std::endl;
+        if (handle_new_entry(dst_dentry_path, dentry, de, new_stx, dir_queue) < 0) {
+          is_error = true;
+          break;
+        }
+      }
+    }
+    if (ceph_closedir(m_src_mnt, new_dirp) < 0) {
+      std::clog << __func__ << ":" << __LINE__ << ": closedir failed for " << ("." + dir_name) << std::endl;
+    }
+    new_dirp = NULL;
+  }
+  if (!is_error) {
+    int rv = 0;
+    std::string dst_snap = m_src_dir + "/"s + m_dst_snap_dir + "/"s + m_new_snap;
+    if ((rv = ceph_mkdir(m_dst_mnt, dst_snap.c_str(), 0755)) < 0 and rv != -EEXIST) {
+      std::clog << __func__ << ":" << __LINE__ << ": failed to create dst snapshot dir " << dst_snap << std::endl;
+      my_errno = -rv;
+      return -1;
+    }
+    return 0;
+  }
+  return -1;
+}
+
+struct timespec SnapReplicator::get_rctime(const std::string& snap_name, const std::string& dir)
+{
+  struct timespec ts{0, 0};
+
+  if (snap_name == ""s)
```

Comment:
when is this case valid?

Reply:
snap_name is an empty string for the very first sync for a directory